source/loader/layers/sanitizer/asan_interceptor.cpp

```diff
@@ -0,0 +1,594 @@
+                        Context, Device, PageSize, &Desc, &PhysicalMem);
+                    if (URes != UR_RESULT_SUCCESS) {
+                        context.logger.error("zePhysicalMemCreate(): {}", URes);
+                        return URes;
+                    }
+                }
+
+                context.logger.debug("zeVirtualMemMap: {} ~ {}",
+                                     (void *)MappedPtr,
+                                     (void *)(MappedPtr + PageSize - 1));
+
+                // FIXME: No flag to check the failed reason is VA is already mapped
+                auto URes = context.urDdiTable.VirtualMem.pfnMap(
+                    Context, (void *)MappedPtr, PageSize, PhysicalMem, 0,
+                    UR_VIRTUAL_MEM_ACCESS_FLAG_READ_WRITE);
+                if (URes != UR_RESULT_SUCCESS) {
+                    context.logger.debug("zeVirtualMemMap(): {}", URes);
+                }
+
+                // Initialize to zero
+                if (URes == UR_RESULT_SUCCESS) {
+                    // Reset PhysicalMem to null since it's been mapped
+                    PhysicalMem = nullptr;
+
+                    const char Pattern[] = {kUnkownRedzoneMagic};
+
+                    auto URes = context.urDdiTable.Enqueue.pfnUSMFill(
+                        Queue, (void *)MappedPtr, 1, Pattern, PageSize,
+                        NumEventsInWaitList, EventsWaitList, Event);
+                    if (URes != UR_RESULT_SUCCESS) {
+                        context.logger.error("urEnqueueUSMFill(): {}", URes);
+                        return URes;
+                    }
+
+                    NumEventsInWaitList = 1;
+                    EventsWaitList = Event;
+                }
+            }
+        }
+
+        const char Pattern[] = {(char)Value};
+        auto URes = context.urDdiTable.Enqueue.pfnUSMFill(
+            Queue, (void *)ShadowBegin, 1, Pattern,
+            (ShadowEnd - ShadowBegin + 1), NumEventsInWaitList, EventsWaitList,
+            Event);
+        if (URes != UR_RESULT_SUCCESS) {
+            context.logger.error("urEnqueueUSMFill(): {}", URes);
+            return URes;
+        }
+    } else {
+        die("Unsupport device type");
+    }
+    return UR_RESULT_SUCCESS;
+}
+
+ur_result_t SanitizerInterceptor::enqueueAllocInfo(
+    ur_context_handle_t Context, ur_device_handle_t Device,
+    ur_queue_handle_t Queue, std::shared_ptr<USMAllocInfo> &AllocInfo,
+    ur_event_handle_t &LastEvent) {
+    // Init zero
+    UR_CALL(enqueueMemSetShadow(Context, Device, Queue, AllocInfo->AllocBegin,
+                                AllocInfo->AllocSize, 0, LastEvent,
+                                &LastEvent));
+
+    uptr TailBegin = RoundUpTo(AllocInfo->UserEnd, ASAN_SHADOW_GRANULARITY);
+    uptr TailEnd = AllocInfo->AllocBegin + AllocInfo->AllocSize;
+
+    // User tail
+    if (TailBegin != AllocInfo->UserEnd) {
+        auto Value = AllocInfo->UserEnd -
```

Comment:
is this specific tail value read somewhere? are is there a layout documentation somewhere?

Reply:
Yes, these values will be read in some functions in libdevice.
This idea is from asan in compiler-rt, I think if someone is familiar with asan, then he/she will know what's the layout of shadow memory of asan.

Reply:
I understand that, but you are relaying on some (hopefully documented) layout/ABI of the compiler-rt. And if that changes, this code will break. That's why I like having source of truth link near soft dependencies such as this one.

Reply:
Makes sense. I've added some comments on memory allocation. Thanks!